server/src/main/scala/services/graphstages/ArrivalsGraphStage.scala

```diff
@@ -260,23 +260,93 @@ class ArrivalsGraphStage(name: String = "",
     }
 
+
+    def scheduleTimeApproximation(searchKey: UniqueArrival, searchKeyArrivalsSourceType: ArrivalsSourceType, foundArrival: Arrival, arrivalsSourceType: ArrivalsSourceType, logMessage: Boolean = false): Option[UniqueArrival] = {
+
+      val arrivalMap: SortedMap[UniqueArrival, Arrival] = arrivalsSourceType match {
+        case LiveArrivals => liveArrivals //Live feeds arrivals
+        case LiveBaseArrivals => liveBaseArrivals // Cirium feeds arrivals
+        case BaseArrivals => forecastBaseArrivals // ACL feeds arrivals
+      }
+
+      val potentialKeyList: Iterable[UniqueArrival] = arrivalMap.keys.collect {
+        case key if key.number == searchKey.number && key.terminal == searchKey.terminal && Math.abs(key.scheduled - searchKey.scheduled) <= 1 * 60 * 60 * 1000 => key
+      }
+
+      val potentialKey = if (potentialKeyList.size > 1) {
+        log.warn(s"Found multiple potentialKey $potentialKeyList for $searchKeyArrivalsSourceType searchKey $searchKey within an hour of scheduled Approximation in sourceType $arrivalsSourceType")
+        potentialKeyList.toSeq.sortBy(_.scheduled).headOption
+      } else {
+        potentialKeyList.headOption
+      }
+
+      potentialKey match {
+        case Some(key) if arrivalMap.get(key).exists(_.Origin == foundArrival.Origin) =>
+          Some(key)
+        case Some(key) => if (logMessage)
+          log.warn(s"ScheduledTime Approximation $searchKeyArrivalsSourceType searchKey $searchKey and origin ${foundArrival.Origin} match not found for potential Key $key in sourceType $arrivalsSourceType")
+          None
+        case None => if (logMessage)
+          log.warn(s"ScheduledTime Approximation search for $searchKeyArrivalsSourceType searchKey $searchKey not found within an hour in sourceType $arrivalsSourceType")
+          None
+      }
+    }
+
     def mergeArrival(key: UniqueArrival): Option[Arrival] = {
       val maybeLiveBaseArrivalWithSanitisedData = liveBaseArrivals.get(key).map(arrivalDataSanitiser.withSaneEstimates)
       val maybeBestArrival: Option[Arrival] = (
         liveArrivals.get(key),
         maybeLiveBaseArrivalWithSanitisedData) match {
-        case (Some(liveArrival), None) => Option(liveArrival)
+
         case (Some(liveArrival), Some(baseLiveArrival)) =>
           val mergedLiveArrival = LiveArrivalsUtil.mergePortFeedWithBase(liveArrival, baseLiveArrival)
           val sanitisedLiveArrival = ArrivalDataSanitiser
             .arrivalDataSanitiserWithoutThresholds
             .withSaneEstimates(mergedLiveArrival)
           Option(sanitisedLiveArrival)
-        case (None, Some(baseLiveArrival)) if forecastBaseArrivals.contains(key) =>
 
-          Option(baseLiveArrival)
+        case (Some(liveArrival), None) if scheduleTimeApproximation(key, LiveArrivals, liveArrival, LiveBaseArrivals, true).isDefined => {
+          val potentialLiveBaseKey = scheduleTimeApproximation(key, LiveArrivals, liveArrival, LiveBaseArrivals)
```

Comment:
We're doing the same thing here in 4 cases so we could move the logic to a function rather than repeating it